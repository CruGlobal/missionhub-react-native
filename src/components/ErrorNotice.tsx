import React from 'react';
import { StyleSheet } from 'react-native';
import { useTranslation } from 'react-i18next';
import { ApolloError } from 'apollo-client';

import theme from '../theme';

import { Text, Touchable } from './common';

const styles = StyleSheet.create({
  errorContainer: {
    backgroundColor: theme.red,
    alignItems: 'center',
    padding: 5,
  },
  white: { color: theme.white },
});

interface ErrorNoticeProps {
  error?: ApolloError;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  refetch: (variables?: any) => Promise<any>;
  message: string;
}

export const ErrorNotice = ({ error, refetch, message }: ErrorNoticeProps) => {
  const { t } = useTranslation('errorNotice');

  return error ? (
    <Touchable style={styles.errorContainer} onPress={refetch}>
      <Text style={styles.white}>{message}</Text>
      {error.networkError && (
        <Text style={styles.white}>{t('networkError')}</Text>
      )}
    </Touchable>
  ) : null;
};
