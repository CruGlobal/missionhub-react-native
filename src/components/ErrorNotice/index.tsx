import React from 'react';
import { View, StyleSheet } from 'react-native';

import { Text } from '../common';
import theme from '../../theme';

const styles = StyleSheet.create({
  container: { backgroundColor: theme.red, alignItems: 'center', padding: 5 },
  message: { color: theme.white },
});

export const ErrorNotice = ({ message }: { message: string }) => (
  <View style={styles.container}>
    <Text style={styles.message}>{message}</Text>
  </View>
);
