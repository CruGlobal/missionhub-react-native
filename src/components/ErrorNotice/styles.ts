import { StyleSheet } from 'react-native';

import theme from '../../theme';

export const styles = StyleSheet.create({
  errorContainer: {
    backgroundColor: theme.darkGrey,
    alignItems: 'center',
    padding: 5,
    flexDirection: 'row',
    justifyContent: 'center',
  },
  text: { ...theme.textRegular16, color: theme.white },
});
