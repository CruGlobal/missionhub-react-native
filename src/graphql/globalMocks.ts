import faker from 'faker/locale/en';

let currentId = 1;
const nextId = () => currentId++;
export const resetGlobalMockSeeds = () => {
  currentId = 1;
  faker.seed(1);
};
resetGlobalMockSeeds();

export const globalMocks = {
  String: () => faker.lorem.words(),
  Int: () => faker.random.number(),
  Float: () => faker.random.number({ precision: 0.01 }),
  Boolean: () => faker.random.boolean(),
  ID: () => nextId(),

  AcceptedChallenge: () => ({
    title: faker.hacker.phrase(),
  }),
  Person: () => ({
    fullName: faker.name.findName(),
  }),
};
