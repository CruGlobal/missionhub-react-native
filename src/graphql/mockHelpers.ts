export const noPages = {
  pageInfo: { hasNextPage: false, hasPreviousPage: false },
};
