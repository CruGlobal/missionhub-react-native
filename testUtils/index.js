import 'react-native';
import renderer from 'react-test-renderer';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// Need to import for all translation screens to work properly
import '../src/i18n';

Enzyme.configure({ adapter: new Adapter() });

export const createMockStore = (state = {}) => {
  return {
    getState: jest.fn(() => (state)),
    dispatch: jest.fn((response) => Promise.resolve(response)),
    subscribe: jest.fn(),
  };
};

export const createMockNavState = (params = {}) => {
  return { state: { params } };
};

export const testSnapshot = (data) => {
  expect(renderer.create(data)).toMatchSnapshot();
};

export const renderShallow = (component, store) => {
  let renderedComponent = shallow(
    component,
    { context: { store: store } }
  );

  // If component has translation wrappers, dive deeper
  while (renderedComponent.is('Translate') || renderedComponent.is('I18n')) {
    renderedComponent = renderedComponent.dive();
  }

  // Render contents of component
  renderedComponent = renderedComponent.dive();
  return renderedComponent;
};

export const testSnapshotShallow = (component, store) => {
  const renderedComponent = renderShallow(component, store);
  expect(renderedComponent).toMatchSnapshot();
  return renderedComponent;
};

export const mockFnWithParams = (obj, method, expectedReturn, ...expectedParams) => {
  return obj[method] = jest.fn(
    (...actualParams) => {
      expect(actualParams).toEqual(expectedParams);
      return expectedReturn;
    }
  );
};
