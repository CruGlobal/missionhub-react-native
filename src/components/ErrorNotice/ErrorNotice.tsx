import React, { ReactNode } from 'react';
import { useTranslation } from 'react-i18next';
import { ApolloError } from 'apollo-client';

import { Text, Touchable, Icon } from '../common';

import { styles } from './styles';

interface ErrorNoticeProps {
  error?: ApolloError;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  refetch: (variables?: any) => Promise<any>;
  message: string;
  specificErrors?: {
    condition: string;
    message: string;
  }[];
}

export const ErrorNotice = ({
  error,
  refetch,
  message,
  specificErrors = [],
}: ErrorNoticeProps) => {
  const { t } = useTranslation('errorNotice');
  const matchedErrors =
    error &&
    specificErrors.reduce((acc: ReactNode[], specificError) => {
      return error.graphQLErrors
        .map(({ message }) => message)
        .includes(specificError.condition)
        ? [
            ...acc,
            <Text key={message} style={styles.text}>
              {specificError.message}
            </Text>,
          ]
        : acc;
    }, []);

  return error ? (
    // Wrapped refetch in function to avoid passing native event to variables argument
    <Touchable style={styles.errorContainer} onPress={() => refetch()}>
      {error.networkError ? (
        <Text style={styles.text}>{t('offline')}</Text>
      ) : matchedErrors && matchedErrors.length > 0 ? (
        matchedErrors
      ) : (
        <Text style={styles.text}>{message}</Text>
      )}
      <Icon name="refresh" />
    </Touchable>
  ) : null;
};
