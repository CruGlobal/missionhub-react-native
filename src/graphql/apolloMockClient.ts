import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-boost';
import { SchemaLink } from 'apollo-link-schema';
import { addMockFunctionsToSchema, IMocks } from 'graphql-tools';
import { buildClientSchema, IntrospectionQuery } from 'graphql/utilities';

import schemaFile from '../../schema.json';

import { globalMocks } from './globalMocks';

export const createApolloMockClient = (mockResolvers: IMocks = {}) => {
  const schema = buildClientSchema(
    (schemaFile as unknown) as IntrospectionQuery,
  );

  addMockFunctionsToSchema({
    schema,
    mocks: { ...globalMocks, ...mockResolvers },
  });

  return new ApolloClient({
    cache: new InMemoryCache(),
    link: new SchemaLink({ schema }),
  });
};
