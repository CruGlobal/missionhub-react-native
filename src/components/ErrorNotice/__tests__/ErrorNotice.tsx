import React from 'react';
import { ApolloError } from 'apollo-client';
import { GraphQLError } from 'graphql';
import { fireEvent } from 'react-native-testing-library';

import { renderWithContext } from '../../../../testUtils';
import { ErrorNotice } from '../ErrorNotice';

// Tests disabled until https://jira.cru.org/browse/MHP-3159 is addressed
xdescribe('ErrorNotice', () => {
  it('should render nothing if no error', () => {
    renderWithContext(
      <ErrorNotice
        message="A test error was fired."
        refetch={() => Promise.resolve()}
      />,
    ).snapshot();
  });
  it('should render an error', () => {
    renderWithContext(
      <ErrorNotice
        message="A test error was fired."
        error={
          new ApolloError({ graphQLErrors: [new GraphQLError('test error')] })
        }
        refetch={() => Promise.resolve()}
      />,
    ).snapshot();
  });
  it('should render a network error', () => {
    renderWithContext(
      <ErrorNotice
        message="A test error was fired."
        error={new ApolloError({ networkError: new Error('test error') })}
        refetch={() => Promise.resolve()}
      />,
    ).snapshot();
  });
  it('should call refetch on press', () => {
    const refetch = jest.fn(() => Promise.resolve());
    const { getByText } = renderWithContext(
      <ErrorNotice
        message="A test error was fired."
        error={
          new ApolloError({ graphQLErrors: [new GraphQLError('test error')] })
        }
        refetch={refetch}
      />,
    );
    fireEvent.press(getByText('A test error was fired.'));
    expect(refetch).toHaveBeenCalled();
  });
});
