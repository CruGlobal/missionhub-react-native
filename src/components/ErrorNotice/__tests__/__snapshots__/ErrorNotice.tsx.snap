// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorNotice should render a network error 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#505256",
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "padding": 5,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#ffffff",
        "fontFamily": "SourceSansPro-Regular",
        "fontSize": 16,
        "fontStyle": "normal",
        "fontWeight": "normal",
        "lineHeight": 24,
      }
    }
  >
    Offline
  </Text>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": undefined,
          "fontSize": 12,
        },
        Array [
          Object {
            "color": "#ffffff",
          },
          Object {
            "fontSize": 18,
          },
          undefined,
        ],
        Object {
          "fontFamily": "Material Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    
  </Text>
</View>
`;

exports[`ErrorNotice should render an a specificError 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#505256",
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "padding": 5,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#ffffff",
        "fontFamily": "SourceSansPro-Regular",
        "fontSize": 16,
        "fontStyle": "normal",
        "fontWeight": "normal",
        "lineHeight": 24,
      }
    }
  >
    A specific error was fired
  </Text>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": undefined,
          "fontSize": 12,
        },
        Array [
          Object {
            "color": "#ffffff",
          },
          Object {
            "fontSize": 18,
          },
          undefined,
        ],
        Object {
          "fontFamily": "Material Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    
  </Text>
</View>
`;

exports[`ErrorNotice should render an error 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#505256",
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "padding": 5,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#ffffff",
        "fontFamily": "SourceSansPro-Regular",
        "fontSize": 16,
        "fontStyle": "normal",
        "fontWeight": "normal",
        "lineHeight": 24,
      }
    }
  >
    A test error was fired.
  </Text>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": undefined,
          "fontSize": 12,
        },
        Array [
          Object {
            "color": "#ffffff",
          },
          Object {
            "fontSize": 18,
          },
          undefined,
        ],
        Object {
          "fontFamily": "Material Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    
  </Text>
</View>
`;

exports[`ErrorNotice should render nothing if no error 1`] = `null`;

exports[`ErrorNotice should show general error if no specific errors are matched 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#505256",
      "flexDirection": "row",
      "justifyContent": "center",
      "opacity": 1,
      "padding": 5,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#ffffff",
        "fontFamily": "SourceSansPro-Regular",
        "fontSize": 16,
        "fontStyle": "normal",
        "fontWeight": "normal",
        "lineHeight": 24,
      }
    }
  >
    A test error was fired.
  </Text>
  <Text
    allowFontScaling={false}
    style={
      Array [
        Object {
          "color": undefined,
          "fontSize": 12,
        },
        Array [
          Object {
            "color": "#ffffff",
          },
          Object {
            "fontSize": 18,
          },
          undefined,
        ],
        Object {
          "fontFamily": "Material Icons",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        Object {},
      ]
    }
  >
    
  </Text>
</View>
`;
